packages/lodestar/src/chain/validation/attestation.ts

```diff
@@ -1,177 +1,286 @@
+  // > Altready check in `verifyHeadBlockAndTargetRoot()`
+
+  const attestationTargetState = await chain.regen.getCheckpointState(attTarget).catch((e) => {
+    throw new AttestationError({code: AttestationErrorCode.MISSING_ATTESTATION_TARGET_STATE, error: e as Error});
+  });
+
+  // [REJECT] The committee index is within the expected range
+  // -- i.e. data.index < get_committee_count_per_slot(state, data.target.epoch)
+  const attIndex = attData.index;
+  const committeeIndices = getCommitteeIndices(attestationTargetState, attSlot, attIndex);
+  const validatorIndex = committeeIndices[bitIndex];
+
+  // [REJECT] The number of aggregation bits matches the committee size
+  // -- i.e. len(attestation.aggregation_bits) == len(get_beacon_committee(state, data.slot, data.index)).
+  // > TODO: Is this necessary? Lighthouse does not do this check
+  if (aggregationBits.length !== committeeIndices.length) {
+    throw new AttestationError({code: AttestationErrorCode.WRONG_NUMBER_OF_AGGREGATION_BITS});
   }
 
-  const expectedSubnet = allForks.computeSubnetForAttestation(attestationTargetState, attestation);
-  if (subnet !== expectedSubnet) {
+  // LH > verify_middle_checks
+  // Run the checks that apply to the indexed attestation before the signature is checked.
+  //   Check correct subnet
+  //   The attestation is the first valid attestation received for the participating validator for the slot, attestation.data.slot.
+
+  // [REJECT] The attestation is for the correct subnet
+  // -- i.e. compute_subnet_for_attestation(committees_per_slot, attestation.data.slot, attestation.data.index) == subnet_id,
+  // where committees_per_slot = get_committee_count_per_slot(state, attestation.data.target.epoch),
+  // which may be pre-computed along with the committee information for the signature check.
+  const expectedSubnet = computeSubnetForSlot(attestationTargetState, attSlot, attIndex);
+  if (subnet !== null && subnet !== expectedSubnet) {
     throw new AttestationError({
       code: AttestationErrorCode.INVALID_SUBNET_ID,
       received: subnet,
       expected: expectedSubnet,
-      job: attestationJob,
     });
   }
 
-  const indexedAttestation = attestationTargetState.getIndexedAttestation(attestation);
-
-  // Do verify signature
-  if (!attestationJob.validSignature) {
-    const signatureSet = allForks.getIndexedAttestationSignatureSet(attestationTargetState, indexedAttestation);
-    if (!(await chain.bls.verifySignatureSets([signatureSet]))) {
-      throw new AttestationError({
-        code: AttestationErrorCode.INVALID_SIGNATURE,
-        job: attestationJob,
-      });
-    }
+  // [IGNORE] There has been no other valid attestation seen on an attestation subnet that has an
+  // identical attestation.data.target.epoch and participating validator index.
+  if (chain.seenAttesters.isKnown(targetEpoch, validatorIndex)) {
```

Comment:
seems like this cache strategy is a great improvement compared to our current one 👍 
confirm that each validator is a unique attester for an epoch as in `computeEpochShuffling`